File: src/games/stendhal/client/TileRenderer.java
/* $Id$ */
/***************************************************************************
 *                      (C) Copyright 2003 - Marauroa                      *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.client;

import java.awt.Graphics;
import java.io.IOException;
import java.io.InputStream;

import org.apache.log4j.Logger;

import games.stendhal.client.sprite.Sprite;
import games.stendhal.client.sprite.Tileset;
import games.stendhal.common.tiled.LayerDefinition;

/**
 * This is a helper class to render coherent tiles based on the tileset. This
 * should be replaced by independent tiles as soon as possible .
 */
class TileRenderer extends LayerRenderer {
	/** the logger instance. */
	private static final Logger logger = Logger.getLogger(TileRenderer.class);
	/** Tileset used for the map data */
	protected Tileset tileset;
	/** Raw map data. Indices of tiles in the tileset. */
	protected int[] map;
	/** The map data converted to tile references */
	protected Sprite[] spriteMap;

	public TileRenderer() {
		map = null;
		spriteMap = null;
	}

	/**
	 * Sets the data that will be rendered.
	 * @param in the stream to read from
	 * @throws IOException
	 *
	 * @throws ClassNotFoundException
	 */
	public void setMapData(final InputStream in) throws IOException,
			ClassNotFoundException {
		final LayerDefinition layer = LayerDefinition.decode(in);
		width = layer.getWidth();
		height = layer.getHeight();

		logger.debug("Layer(" + layer.getName() + "): " + width + "x" + height);

		map = layer.expose();
	}

	/**
	 * Set the tileset.
	 *
	 * @param tileset
	 *            The tileset.
	 */
	@Override
	public void setTileset(final Tileset tileset) {
		this.tileset = tileset;
	}

	/**
	 * Initialize the sprite map from the tileset and the map data.
	 *
	 * @return true if the map is ready to be used, false otherwise.
	 */
	private boolean initSpriteMap() {
		if (spriteMap == null) {
			if (tileset != null) {
				/*
				 * Cache sprites
				 */
				spriteMap = new Sprite[map.length];

				int i = spriteMap.length;

				while (i-- != 0) {
					spriteMap[i] = tileset.getSprite(map[i]);
				}
			} else {
				return false;
			}
		}
		return true;
	}

	@Override
	public void draw(Graphics g, int x, int y, final int w, final int h) {
		if (!initSpriteMap()) {
			return;
		}

		final int endX = Math.min(x + w, getWidth());
		final int endY= Math.min(y + h, getHeight());

		int sy = y * IGameScreen.SIZE_UNIT_PIXELS;
		for (int j = y; j < endY; j++) {
			int mapidx = (j * width) + x;
			int sx = x * IGameScreen.SIZE_UNIT_PIXELS;

			for (int i = x; i < endX; i++) {
				spriteMap[mapidx].draw(g, sx, sy);
				mapidx++;
				sx += IGameScreen.SIZE_UNIT_PIXELS;
			}
			sy += IGameScreen.SIZE_UNIT_PIXELS;
		}
	}
}

File: srcjs/stendhal/renderer/TileRenderer.ts
/***************************************************************************
 *                 (C) Copyright 2003-2021 Faiumoni e. V.                  *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { Sprite } from "sprite/Sprite";
import { Tileset } from "sprite/Tileset";
import { LayerRenderer } from "./LayerRenderer";

 
/**
 * This is a helper class to render coherent tiles based on the tileset. This
 * should be replaced by independent tiles as soon as possible .
 */
class TileRenderer extends LayerRenderer {

    /** Tileset used for the map data */
    private tileset?: Tileset;

    /** Raw map data. Indices of tiles in the tileset. */
    private map?: number[];

    /** The map data converted to tile references */
    private spriteMap?: Sprite[];

    /**
     * Sets the data that will be rendered.
     * @param in the stream to read from
     * @throws IOException
     *
     * @throws ClassNotFoundException
     */
    /* TODO
    public setMapData(layer: LayerDefinition): void {
        this.width = layer.getWidth();
        this.height = layer.getHeight();
        this.map = layer.expose();
    }*/

    /**
     * Set the tileset.
     *
     * @param tileset
     *            The tileset.
     */
    public setTileset(tileset: Tileset): void {
        this.tileset = tileset;
    }

    /**
     * Initialize the sprite map from the tileset and the map data.
     *
     * @return true if the map is ready to be used, false otherwise.
     */
    private initSpriteMap(): boolean {
        if (this.spriteMap) {
            return true;
        }

        if (!this.tileset || !this.map) {
            return false;
        }
            
        // Cache sprites
        this.spriteMap = new Array<Sprite>(this.map.length);

        for (let i = this.spriteMap.length; i > 0; i--) {
            this.spriteMap[i] = this.tileset.getSprite(this.map[i]);
        }
        return true;
    }

    public draw(g: CanvasRenderingContext2D, x: number, y: number, w: number, h: number): void {
        if (!this.initSpriteMap() || !this.spriteMap) {
            return;
        }

        const SIZE_UNIT_PIXELS = 32;  // TODO this is defined in IGameScreen
        const endX = Math.min(x + w, this.getWidth());
        const endY = Math.min(y + h, this.getHeight());

        let sy = y * SIZE_UNIT_PIXELS;
        for (let j = y; j < endY; j++) {
            let mapidx = (j * this.width) + x;
            let sx = x * SIZE_UNIT_PIXELS;

            for (let i = x; i < endX; i++) {
                this.spriteMap[mapidx].draw(g, sx, sy);
                mapidx++;
                sx += SIZE_UNIT_PIXELS;
            }
            sy += SIZE_UNIT_PIXELS;
        }
    }
}
